File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mp = { ")": "(", "]": "[", "}":"{"}
        stack = []
        for c in s:
            if c not in mp:
                stack.append(c)
                continue
            if not stack or stack[-1] != mp[c]:
                return False
            stack.pop()
        return not stack

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> p = {
            {')', '('},
            {']', '['},
            {'}', '{'},
        };
        stack<int> open;

        for (auto c: s) {
            if (p.find(c) != p.end()) {
                // found closing bracket
                if (open.empty()) return false;
                if (open.top() != p[c]) return false;
                open.pop();
            } else open.push(c);
        }

        return open.empty();
    }
};